src/batchUploader.ts

```diff
@@ -129,21 +131,68 @@ export class BatchUploader {
         return offlineStoragePercentage >= rampNumber;
     }
 
+    private shouldDebounceAST(): boolean {
+        const now = Date.now();
+        if (now - this.lastASTEventTime < this.AST_DEBOUNCE_MS) {
+            return true;
+        }
+
+        this.lastASTEventTime = now;
+
+        return false;
+    }
+
+    private createBackgroundASTEvent(): SDKEvent {
+        const now = Date.now();
+        const { _Store } = this.mpInstance;
+        return {
+            EventDataType: MessageType.AppStateTransition,
+            EventName: 'Application Exit',
+            EventCategory: EventType.Other,
+            Timestamp: now,
+            SessionId: _Store.sessionId,
+            MPID: this.mpInstance.Identity.getCurrentUser()?.getMPID(),
+            DeviceId: _Store.deviceId,
+            IsFirstRun: false,
+            SourceMessageId: '',
+            SDKVersion: Constants.sdkVersion,
+            CustomFlags: {},
+            UserAttributes: {},
+            UserIdentities: [],
+            SessionStartDate: _Store.sessionStartDate?.getTime() || now,
+            Debug: _Store.SDKConfig.isDevelopmentMode,
+            CurrencyCode: null,
+            ExpandedEventCount: 0,
+            ActiveTimeOnSite: this.mpInstance._timeOnSiteTimer?.getTimeInForeground() || 0,
+            IsBackgroundAST: true
+        };
+    }
+
     // Adds listeners to be used trigger Navigator.sendBeacon if the browser
     // loses focus for any reason, such as closing browser tab or minimizing window
     private addEventListeners() {
         const _this = this;
 
+        const handleExit = () => {
+            // Check for debounce before creating and queueing event
+            if (_this.shouldDebounceAST()) {
+                return;
+            }
+            // Add application state transition event to queue
+            const event = _this.createBackgroundASTEvent();
+            _this.queueEvent(event);
+            // Then trigger the upload with beacon
+            _this.prepareAndUpload(false, _this.isBeaconAvailable());
+        };
+
         // visibility change is a document property, not window
         document.addEventListener('visibilitychange', () => {
-            _this.prepareAndUpload(false, _this.isBeaconAvailable());
-        });
-        window.addEventListener('beforeunload', () => {
-            _this.prepareAndUpload(false, _this.isBeaconAvailable());
-        });
-        window.addEventListener('pagehide', () => {
-            _this.prepareAndUpload(false, _this.isBeaconAvailable());
+            if (document.visibilityState === 'hidden') {
```

Comment:
Nitpick: I get the appeal to reduce the footprint of this listener, but since `handleExit()` isn't being reused at the moment, I don't see a point in making it a convenience method. It also serves to highlight that this specific listener is doing something different and prevents the need to dig too deep for what's going on.

Reply:
i'm not sure i understand this comment.  handleExist is being used 3x in this file.  so that's why it should be its own method to be DRY

Reply:
Disregard this comment. I didn't realize at the time it was being used in multiple places.